packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -1756,6 +1768,592 @@ export class ProjectService extends BaseService {
+                    span.setStatus({
+                        code: 2, // ERROR
+                        message: getErrorMessage(e),
+                    });
+                    throw e;
+                } finally {
+                    span.end();
+                }
+            },
+        );
+    }
+
+    async runPaginatedMetricQuery({
+        user,
+        projectUuid,
+        dateZoomGranularity,
+        context = QueryExecutionContext.API,
+        metricQuery,
+        csvLimit,
+        page,
+        pageSize,
+    }: PaginateMetricQueryArgs): Promise<ApiPaginatedQueryResults> {
+        if (!isUserWithOrg(user)) {
+            throw new ForbiddenError('User is not part of an organization');
+        }
+        const { organizationUuid } =
+            await this.projectModel.getWithSensitiveFields(projectUuid);
+
+        if (
+            user.ability.cannot(
+                'manage',
+                subject('Explore', { organizationUuid, projectUuid }),
+            )
+        ) {
+            throw new ForbiddenError();
+        }
+
+        if (
+            metricQuery.customDimensions?.some(isCustomSqlDimension) &&
+            user.ability.cannot(
+                'manage',
+                subject('CustomSql', { organizationUuid, projectUuid }),
+            )
+        ) {
+            throw new ForbiddenError(
+                'User cannot run queries with custom SQL dimensions',
+            );
+        }
+
+        const queryTags: RunQueryTags = {
+            organization_uuid: organizationUuid,
+            project_uuid: projectUuid,
+            user_uuid: user.userUuid,
+            explore_name: metricQuery.exploreName,
+            query_context: context,
+        };
+
+        return this.runPaginatedQuery(
+            {
+                user,
+                metricQuery,
+                projectUuid,
+                exploreName: metricQuery.exploreName,
+                csvLimit,
+                context: QueryExecutionContext.EXPLORE,
+                queryTags,
+                granularity: dateZoomGranularity,
+                page,
+                pageSize,
+            },
+            formatRow,
+        );
+    }
+
+    async runPaginatedQueryIdQuery({
+        user,
+        projectUuid,
+        context = QueryExecutionContext.API,
+        exploreName,
+        ...rest
+    }: PaginateQueryIdArgs): Promise<ApiPaginatedQueryResults> {
+        if (!isUserWithOrg(user)) {
+            throw new ForbiddenError('User is not part of an organization');
+        }
+        const { organizationUuid } =
+            await this.projectModel.getWithSensitiveFields(projectUuid);
+
+        if (
+            user.ability.cannot(
+                'manage',
+                subject('Explore', { organizationUuid, projectUuid }),
+            )
+        ) {
+            throw new ForbiddenError();
+        }
+
+        // TODO paginate: we need metadata to be able enforce this permission
```

Comment:
We need to change the `manage` Explore check for this to work for viewers. And we only validate the query content on creation (aka runPaginatedMetricQuery )

We should validate:
- query project uuid match URL project uuid
- user that created query is the same as user authenticated 

These are only possible once we store the metadata. For now, we can check if the user can view project uuid.